src/widgets/input.ts

```diff
@@ -14,30 +22,46 @@ const enum DataTypes {
 export class Input extends Widget {
 	private readonly VALID_PROPERTIES = ['id', 'editable', 'minlength', 'maxlength', 'required', 'format']
 
-	constructor(props: any, name: string) {
+	constructor(props: any, name: string, private _customDefinition?: WidgetConfig) {
 		super(props, name)
 	}
 
 	public renderToString(): string {
+		const input = this.createInput()
+		const label = createLabel(this.name, this.props.id.toString())
+		return this.wrap(label + input)
+	}
+
+	private createInput(): string {
```

Comment:
This approach is not effective anymore. All the widgets' definition are now loaded from the config file. Every widget tag can contain a variable where the properties are placed. The rendering becomes generic, and can be implemented in a single class. The widgets need only to interpret the properties and to render them, not the widget itself.